File: windows/PetRunner.Core/Physics.cs
namespace PetRunner.Core;

public struct MotionState(double x, double y, double velocityX, double velocityY)
{
    public double X = x;
    public double Y = y;
    public double VelocityX = velocityX;
    public double VelocityY = velocityY;
}

public readonly record struct BounceResult(bool Horizontal, bool Vertical);

public static class PhysicsEngine
{
    public static BounceResult Step(
        ref MotionState motion,
        SizeD size,
        RectD bounds,
        double deltaTime,
        double velocityRetentionPerSecond = 0.18,
        double restitution = 0.72,
        double stopSpeed = 8,
        double maximumDeltaTime = 1)
    {
        var dt = Math.Clamp(deltaTime, 0, maximumDeltaTime);
        if (dt <= 0) return default;

        motion.X += motion.VelocityX * dt;
        motion.Y += motion.VelocityY * dt;
        var maxX = Math.Max(bounds.X, bounds.X + bounds.Width - size.Width);
        var maxY = Math.Max(bounds.Y, bounds.Y + bounds.Height - size.Height);
        var horizontal = false;
        var vertical = false;

        if (motion.X < bounds.X)
        {
            motion.X = bounds.X;
            motion.VelocityX = Math.Abs(motion.VelocityX) * restitution;
            horizontal = true;
        }
        else if (motion.X > maxX)
        {
            motion.X = maxX;
            motion.VelocityX = -Math.Abs(motion.VelocityX) * restitution;
            horizontal = true;
        }
        if (motion.Y < bounds.Y)
        {
            motion.Y = bounds.Y;
            motion.VelocityY = Math.Abs(motion.VelocityY) * restitution;
            vertical = true;
        }
        else if (motion.Y > maxY)
        {
            motion.Y = maxY;
            motion.VelocityY = -Math.Abs(motion.VelocityY) * restitution;
            vertical = true;
        }

        var retention = Math.Pow(velocityRetentionPerSecond, dt);
        motion.VelocityX *= retention;
        motion.VelocityY *= retention;
        if (Math.Sqrt(
                motion.VelocityX * motion.VelocityX + motion.VelocityY * motion.VelocityY) < stopSpeed)
        {
            motion.VelocityX = 0;
            motion.VelocityY = 0;
        }
        return new BounceResult(horizontal, vertical);
    }

    public static (double X, double Y) Clamp(double x, double y, SizeD size, RectD bounds) =>
        (
            Math.Clamp(x, bounds.X, Math.Max(bounds.X, bounds.X + bounds.Width - size.Width)),
            Math.Clamp(y, bounds.Y, Math.Max(bounds.Y, bounds.Y + bounds.Height - size.Height))
        );

    public static (double X, double Y) CenteredOrigin(SizeD size, RectD bounds)
    {
        var origin = (bounds.X + (bounds.Width - size.Width) / 2, bounds.Y + (bounds.Height - size.Height) / 2);
        return Clamp(origin.Item1, origin.Item2, size, bounds);
    }
}

File: rust/crates/petrunner-core/src/physics.rs
#[derive(Clone, Copy, Debug, PartialEq)]
#[repr(C)]
pub struct MotionState {
    pub x: f64,
    pub y: f64,
    pub velocity_x: f64,
    pub velocity_y: f64,
}

#[derive(Clone, Copy, Debug, PartialEq)]
#[repr(C)]
pub struct Size {
    pub width: f64,
    pub height: f64,
}

#[derive(Clone, Copy, Debug, PartialEq)]
#[repr(C)]
pub struct Rect {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct PhysicsConfig {
    pub velocity_retention_per_second: f64,
    pub restitution: f64,
    pub stop_speed: f64,
    pub maximum_delta_time: f64,
}

impl Default for PhysicsConfig {
    fn default() -> Self {
        Self {
            velocity_retention_per_second: 0.18,
            restitution: 0.72,
            stop_speed: 8.0,
            maximum_delta_time: 1.0,
        }
    }
}

impl PhysicsConfig {
    pub fn step(
        self,
        motion: &mut MotionState,
        size: Size,
        bounds: Rect,
        delta_time: f64,
    ) -> (bool, bool) {
        let dt = delta_time.clamp(0.0, self.maximum_delta_time);
        if dt <= 0.0 || !dt.is_finite() {
            return (false, false);
        }
        motion.x += motion.velocity_x * dt;
        motion.y += motion.velocity_y * dt;
        let max_x = bounds.x.max(bounds.x + bounds.width - size.width);
        let max_y = bounds.y.max(bounds.y + bounds.height - size.height);
        let mut horizontal = false;
        let mut vertical = false;
        if motion.x < bounds.x {
            motion.x = bounds.x;
            motion.velocity_x = motion.velocity_x.abs() * self.restitution;
            horizontal = true;
        } else if motion.x > max_x {
            motion.x = max_x;
            motion.velocity_x = -motion.velocity_x.abs() * self.restitution;
            horizontal = true;
        }
        if motion.y < bounds.y {
            motion.y = bounds.y;
            motion.velocity_y = motion.velocity_y.abs() * self.restitution;
            vertical = true;
        } else if motion.y > max_y {
            motion.y = max_y;
            motion.velocity_y = -motion.velocity_y.abs() * self.restitution;
            vertical = true;
        }
        let retention = self.velocity_retention_per_second.powf(dt);
        motion.velocity_x *= retention;
        motion.velocity_y *= retention;
        if motion.velocity_x.hypot(motion.velocity_y) < self.stop_speed {
            motion.velocity_x = 0.0;
            motion.velocity_y = 0.0;
        }
        (horizontal, vertical)
    }

    #[must_use]
    pub fn clamped_origin(origin_x: f64, origin_y: f64, size: Size, bounds: Rect) -> (f64, f64) {
        let max_x = bounds.x.max(bounds.x + bounds.width - size.width);
        let max_y = bounds.y.max(bounds.y + bounds.height - size.height);
        (
            origin_x.clamp(bounds.x, max_x),
            origin_y.clamp(bounds.y, max_y),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clamps_bounces_and_stops() {
        let mut motion = MotionState {
            x: 95.0,
            y: 40.0,
            velocity_x: 200.0,
            velocity_y: 0.0,
        };
        let (horizontal, _) = PhysicsConfig {
            velocity_retention_per_second: 1.0,
            stop_speed: 0.0,
            ..PhysicsConfig::default()
        }
        .step(
            &mut motion,
            Size {
                width: 10.0,
                height: 10.0,
            },
            Rect {
                x: 0.0,
                y: 0.0,
                width: 100.0,
                height: 100.0,
            },
            0.1,
        );
        assert!(horizontal);
        assert_eq!(motion.x, 90.0);
        assert!(motion.velocity_x < 0.0);
    }
}
